pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java

```diff
@@ -632,25 +632,29 @@ protected void internalDeletePartitionedTopic(AsyncResponse asyncResponse, boole
                     return;
                 }
             }
+            // Only tries to delete the authentication policies for partitioned topic when all its partitions are
+            // successfully deleted
+            pulsar().getBrokerService().deleteTopicAuthenticationWithRetry(topicName.toString());
```

Comment:
`NamespaceResources#setPoliciesAsync()` is an atomic update to zookeeper with optimistic lock, so maybe this is already an atomic update?

Besides, I use retry logic here (as it's typical to generate `BadVersionException` here), so maybe this is an at-least-once semantics?

As for the `asyncResponse`, I used to think that this deletion operation shouldn't  affect the response because it's not the main goal of the method. Do you mean that when the deletion of authentication policies fails, the whole method should throw an exception?

Reply:
Thanks for clarifying. I should have been more specific. I was referencing atomicity for both this update and the `deletePartitionedTopicAsync` update. My only concern is that failure could lead to corrupt state. I am not familiar enough with Pulsar's interactions with ZK to speak authoritatively here, which is why I asked @eolivelli to take a look.

Reply:
Got it. Thanks for your comments.

Reply:
Maybe we need to add the delete topic authentication policies in the delete topic async operation chain?

Reply:
@gaoran10 I think ok here, If users want to delete a topic, but it failed. users should try to delete it again. But if delete the topic succeed, delete the authentication failed, the authentication will be permanently retained in Zookeeper

Reply:
Maybe we need to ensure all resources of the topic all clean up, then delete the topic from the ManagedLedger?

Reply:
> Maybe we need to ensure all resources of the topic all clean up, then delete the topic from the ManagedLedger? It seems that the result of the method `deleteTopicAuthenticationWithRetry` didn't affect the delete topic operation result.

For partitioned topic, the deletion of ManagedLedger is in `pulsar().getAdminClient().topics().deleteAsync(topicNamePartition.toString(), force)`, which is before the deletion of metadata and authentication policies. Do you mean that we should modify the logic here?